Filter streams on subscriptions page (resolves #565)

mjec opened the pull request:
This is additional to PR #684. It provides filtering on the subscriptions page.

Where the user has `can_create_streams` permission (i.e. always at the moment), filtering is applied when the user types into the create stream box. I think this is the key use case (and avoids creating similar but not precisely-duplicate streams).

If the user lacks that permission, a specific filtering box is rendered instead.

smarx commented:
*Automated message from Dropbox CLA bot*

@mjec, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
Awesome, thanks for working on this!  Looks like some lint failures; I'll take a look at the actual code tomorrow.

I should note that this conflicts from a design perspective with https://github.com/zulip/zulip/pull/514; we'll need to figure out what strategy we want to take for the overall experience.  What do you think is the right design for this @mjec and @disissid?  @mjec it might be helpful if you posted a screenshot of this UI to help with the discussion.

@wdaher in case he has thoughts.

mjec commented:
I completely missed #514; sorry!

I have no strong feelings one way or another on design. I do think it's important to have a way to search streams you are not subscribed to.

Attached are screenshots of the filter box when `can_create_streams` is false. Perhaps adding this box, say left-aligned, and the create new stream box right-aligned, but otherwise in accordance with #514?

![filter-in-settings-inactive](https://cloud.githubusercontent.com/assets/556748/14842732/291090ca-0c18-11e6-9047-10d9a9bd618d.png)

![filter-in-settings-active](https://cloud.githubusercontent.com/assets/556748/14842737/2cad08a8-0c18-11e6-9d92-74a2ddc8d7b6.png)

timabbott commented:
@mjec sorry for the slow reply here!  Here's a path forward for merging this PR and #514:
* First, you can squash the little fixup commits into the commit that introduced the bug -- that's how we do things on the Zulip project, to have a nice clean history.
* Second, I think a reasonable approach for integrating this with #514 is for you to just add the current #514 code as a commit inside your PR, and make the design work reasonably (I think maybe just making your filter UI be left-justified and the create stream button be right-justified would work as a reasonable approach).
* It looks like @disissid has moved on to other projects, so you'll probably need to fix the little bugs that are open in my code review of #514; I think that was pretty close to being done, so this should be fairly straightforward.

If you're game for finishing that combined project, it'd be super awesome!  If you don't have time for that, you should just clean this up to be nicely mergable (aka the "First" task above) and I can figure out a way to deal with #514 later...

mjec commented:
Thanks @timabbott, I'm happy to combine with #514 as suggested. I'm not sure I'll have time before Monday but I'll do it then.

timabbott commented:
OK, definitely no worries on timeline -- I'm going to be on vacation next week so probably won't review until the week after regardless.

mjec commented:
@timabbott I finally got around to doing this. Squashed into two commits, and rebased off current master. Let me know if anything further needs doing.

timabbott commented:
Nice!  I'll try to find time to look today

mjec commented:
Thanks @timabbott, though you'd be forgiven if it took you a bit more time, given it took me a month ;)

timabbott commented on memcached_prefix:
```diff
@@ -0,0 +1 @@
+689f5ba0a0bab644aaa29626c326d455:
```
this shouldn't be being committed (it's in .gitignore; I'm a bit surprised this was possible at all).

timabbott commented:
Well, 4 days is less good than I'd hoped.  I played with this a bit and found a few issues:
* If you type some stuff into the filter box and hit enter, it'll pop up the stream creation form, whereas it probably should expand the top stream and scroll you to it (or something).

timabbott commented on static/styles/zulip.css:
```diff
@@ -2696,13 +2696,22 @@ button.topic_edit_cancel {
+
+
+#listAllStreams.hideAllStreams{
+    display: block;
```
Is this CSS used?

timabbott commented on static/js/subs.js:
```diff
@@ -698,16 +704,9 @@ $(function () {
-            ajaxSubscribe(stream);
-        }
+            show_new_stream_modal();
+            }
```
this indentation change is wrong

timabbott commented on frontend_tests/casper_tests/04-subscriptions.js:
```diff
@@ -12,13 +12,14 @@ casper.then(function () {
 });
-casper.waitForText('Waseemio', function () {
-    casper.test.assertTextExists('Create stream Waseemio', 'Modal for specifying new stream users');
+casper.waitForText('Create stream', function () {
```
Can we replace this with a `waitForSelector`?  Those are generally less fragile...

timabbott commented on static/js/subs.js:
```diff
@@ -743,19 +742,27 @@ $(function () {
-            principals,
-            $('#stream_creation_form input[name=privacy]:checked').val() === "invite-only",
-            $('#announce-new-stream input').prop('checked')
+        var stream_status = compose.check_stream_existence(stream);
```
We shouldn't be using this **synchronous** check for whether the stream exists; sync queries are problematic and there's a race here.  If we want to do something different in the event that the stream already exists, we should pass an argument to the subscribe call here indicating we want to create a new stream and it should throw an error if the stream already exists (the backend doesn't support this, I think, but it'd be easy to add).

timabbott commented on memcached_prefix:
```diff
@@ -1 +0,0 @@
-689f5ba0a0bab644aaa29626c326d455:
```
oh, I see, you fixed memcached_prefix having been committed, but in the second commit, so it gets added and then removed in this branch

timabbott commented on templates/zerver/stream_creation_prompt.html:
```diff
@@ -7,9 +7,9 @@ <h3 id="stream-creation-label">{% trans %}Create stream{% endtrans %}</h3>
+            <b>{% trans %}Stream name{% endtrans %}</b><br />
             <input type="text" name="stream_name" id="create_stream_name"
-                 placeholder="Stream name" value="" />
+                 placeholder="{{ _('Stream name') }}" value="" />
```
you should move these fixes to the first commit into the first commit; otherwise the first commit will totally fail tests when we're bisecting.

timabbott commented:
Gave this a read and posted a few more comments on the code; most of the issues are things of the form "the first commit does something wrong, and the second fixes it", where you need to rebase to move those fixes into the first commit.  Check out the Zulip coding style guidelines here for more details: https://zulip.readthedocs.org/en/latest/code-style.html#version-control

Thanks for working on this @mjec!  I'm really excited about getting these improvements in.

timabbott commented:
(One more note: you should probably rebase this branch on top of master, since it's a ways behind now)

mjec commented:
@timabbott thanks for this! Rebased as requested; I think this should address all the points you raised.

In particular:

* If you type some stuff into the filter box and hit enter, this will expand or contract the top stream; otherwise it will open the stream creation box. I didn't have it scroll to because the stream is already always at the top.

* Search box is now aligned with the stream names.

timabbott commented:
Thanks for working on htis @mjec!  I did some more testing of this, and noticed a few bugs:
* The first commit fails the Casper tests, so we can't merge just that one.
* The first commit doesn't give a clear error message when you try to create a stream that already exists (it just closes the modal, losing whatever you'd typed in for what you want to call things).  Probably the best behavior we could do here is do a check in JS as soon as you finish typing a stream name for whether that stream already exists; if it does, show an error then (can be async).  And regardless we certainly want to show the error we get from the backend if the user clicks "create" anyway (I think we might need to change the backend endpoint to support taking a `require_create` argument to make it throw an error in this case).  
* With the second commit, if you tab over to the "create stream" button with the keyboard, and hit enter, it still expands the first stream in the search, whereas the user probably expects it to open the create stream modal.
* The `should_list_all_streams` functionality is somewhat broken.  You can test this by hardcoding that value to `false`.  The (perhaps ill-conceived) intent of that feature is to mimic how Zephyr works, in that stream names are not public information, and there's also no announcing streams or whatnot (all streams already exist by default).  So for that case, we basically want a way to enter a stream name and hit enter and be subscribed (you can check out how this works on master to see what I mean).  While that feature is arguably somewhat ill-conceived, it'd be good to preserve that feature working correctly.

timabbott commented:
Tagging this as working on #563

timabbott commented:
@mjec how's this coming along?  Let me know if there's anything I can do to help; would love to get this finished :)

brainwane commented:
I believe @trueskawka is working on updating & resubmitting this.

brainwane commented:
Alicja has rebased @mjec's code, fixed failing tests, added one missing test and rewritten the functionality according to Tim's comments on the original PR, and mentioned that she was going to create a new PR today. 
She asked:

> One functionality question at this point - should "create new stream" button work without a stream name on input (since the modal was changed as compared to the current functionality)? I've assumed yes and implemented this option, but maybe I was too fast to assume that.

My opinion: yes.

timabbott commented:
Agreed, the "create new stream" button should work without a stream name input.
